feat(workflow_engine): Process Workflows

saponifi3d opened the pull request:
## Description
Get to the point where process_workflows gets the list of triggered workflows. At this point we'll know the list of actions to filter / evaluate.

Next steps will be to add `processors/action.py` which will have a method that takes the list of triggered workflows, evaluates the filters on an action. Most of this work is encapsulated inside of the DataCondition.

saponifi3d committed:
Scaffold for process_workflows

saponifi3d committed:
Get the detector lookups working correctly from evidence data. Works for error events / occurrences

saponifi3d committed:
Create a workflow with a trigger condition, get to the point where it's triggered.

saponifi3d committed:
add tests to validate we are to the triggered workflows portion. improved the nested value comparison by updating the type and adding a handler for it to filter the value input. its a little cleaner?

saponifi3d committed:
Fixing the workflow test factory, i think this likely got lost on another PR from merge conflicts

saponifi3d committed:
make a base test class that can create all the different models a setup situations for us. upate the model / processor tests to use this base test

saponifi3d commented on src/sentry/workflow_engine/models/data_condition.py:
```diff
@@ -74,32 +95,27 @@ def get_condition_result(self) -> DataConditionResult:
-                "Invalid comparison value", extra={"comparison": self.comparison, "id": self.id}
-            )
-            return None
+        if self.type in FILTER_VALUE_BY_CONDITIONS:
```
🤔 it might be nicer to have a the `DataCondition` base class, then have another class that extends this; `GroupEventDataCondition` - that is always passed a `GroupEvent` as the `value` in evaluate - that would allow us to be explicit about some of the mappings setup ehre.

saponifi3d committed:
remove testutils folder, didn't add much

saponifi3d committed:
Add some comments about what's next, starting to think about opening PR for this chunk

cathteng commented on src/sentry/workflow_engine/processors/workflow.py:
```diff
@@ -0,0 +1,46 @@
+    return triggered_workflows
+
+
+def process_workflows(evt: GroupEvent):
```
can we pass additional information alongside the event here in a dict? there is some information outside of the event that is calculated and passed along that would be evaluated in DataConditions

https://github.com/getsentry/sentry/blob/a53b18fab531059d1e94dfcdfdd5f9698f9d93a7/src/sentry/tasks/post_process.py#L997-L1002

but if we can add it to the GroupEvent itself, we wouldn't need to pass the extra info dict

saponifi3d replied:
we can probably pass the job to here rather than the group_event if needed, but i think some of these also feel like replicated data from the group_event model itself. for example, how is `is_new` different from `group_event.group.times_seen == 1`.

I think the best case scenario is that we _don't_ pass the job though, i think it's generally better if we can understand what's happening from the group event model itself, since that's what the detector is creating in the end.

ceorourke replied:
Some of these are fairly complicated like `is_regression` is calculated [here](https://github.com/getsentry/sentry/blob/master/src/sentry/event_manager.py#L1638) and it'd be nice to not have to recalculate if it can be passed in.

saponifi3d replied:
👍 there's a thread with snigda that discusses how we might be able to align on just using the group event: https://sentry.slack.com/archives/C07H2DHMSS0/p1733177862010779

worst case, this could be a fairly easy update to take in a `PostProcessJob` and just get the group event from that. I'm hoping to keep this as just the GroupEvent if possible to make this method a bit more reusable. (we can also add another processing method in post_process that invokes process_workflows).

i'm figuring we kick the can down the road a little on this in particular since the workflow_engine platform just needs the occurrence right now.

saponifi3d committed:
feat(workflow_engine): DataCondition handler registry (#81356)

## Description
Adds a registry of DataConditionHandlers. This allows us to trigger
different handlers (that are fully typed). Continues to support
operators as well, so we can handle any kind of logic here.

saponifi3d committed:
Fix tests now that things are more type safe

saponifi3d committed:
Update the comparison info to validate that it raises errors now

cathteng commented on src/sentry/workflow_engine/models/data_condition.py:
```diff
@@ -74,32 +85,44 @@ def get_condition_result(self) -> DataConditionResult:
+    def get_condition_handler(self) -> DataConditionHandler[T] | None:
         try:
-            condition = Condition(self.condition)
+            condition_type = Condition(self.type)
```
should we use the `Condition` StrEnum as the possible choices for `DataCondition.type`?

saponifi3d replied:
yeah - i was thinking we'll probably want to change the schema around, but it's kinda hard to do so  punting on that for now. right now it's weird cause `condition` is used for `gt`, `lt`, etc but `type` is used for the handler stuff. i'm thinking the `type` field should use the enum, then figure out something to do with the `condition` field (i'm using it to filter / pass configuration info for custom handlers - so maybe the json config base? not sure, so i'm pushing along in other areas until that's a bit more clear since it's so hard to do migrations. happy to take recommendations!!)

saponifi3d committed:
simplify process_workflows a bit more

ceorourke commented on tests/sentry/workflow_engine/test_base.py:
```diff
@@ -0,0 +1,114 @@
+
+    def create_detector_and_workflow(
+        self,
+        name_prefix="test",
```
Can this be inferred from the `detector_type`? I see in `TestProcessWorkflows` it's passing `"error"` as the `name_prefix` when it passes in the errors detector type, and for the issue occurrence one it doesn't pass either. Is there a time when we'd be passing something else?

saponifi3d replied:
ah, yes, i'll remove. (i was using it for debugging and failed to clean it up, ty!)

saponifi3d replied:
i just did this as a default name in tests, so i can identify them uniquely / easily, but wanted a default so i only have to add when i have multiple detectors.

if we based it on `detector_type` and had two detectors of the same time in a test, it might be a bit confusing.

ceorourke reviewed (approved):
generally lgtm, the things discussed can be added later on if the goal here is just to get the basic thing working

cathteng commented on src/sentry/workflow_engine/models/workflow.py:
```diff
@@ -47,3 +49,14 @@ class Meta:
+    def evaluate_trigger_conditions(self, evt: GroupEvent) -> bool:
+        """
+        Evaluate the conditions for the workflow trigger and return the results.
+        If there isn't a when_condition_group, the workflow should always trigger.
```
it should only trigger if the IF conditions pass too, right? will they be evaluated in this function?

saponifi3d replied:
yep! i'll update the comment to make it clearer.

saponifi3d committed:
Add improvements to the AI comment, but these are still mostly AI
